README.md
# bem-templates-react-ts-v1

Templates set for using with [bem create command](https://github.com/bem-tools/bem-tools-create "bem-tools/bem-tools-create: bem create command") in bem-classic (i-bem/bem-xjst) projects


CHANGELOG.md
# CHANGELOG

- 2018.11.28, 02:16 -- Moved (duped) to github.


utils.js
/* eslint-env es6, node, commonjs */
/**
 * @module template utils
 * @author lilliputten <lilliputten@yandex.ru>
 * @since 2018.09.29, 23:06
 * @version 2018.10.25, 03:43
 */

const fs = require('fs');

// Use `config.js` if you need another from default (`.config.default.js`) parameters.
let configFile = './.config.js';
if (!fs.existsSync(configFile)) {
  configFile = './.config.default.js';
}
const config = require(configFile);

module.exports = Object.assign({

  /** capitalize ** {{{
   */
  capitalize: function(v) {
    return v ? ( String(v)[0].toUpperCase() + String(v).substr(1, v.length) ) : '';
  },/*}}}*/

  /** getModValString ** {{{
   */
  getModValString: function(v) {
    return v == null || v === true ? '' : v;
  },/*}}}*/
  /** hasNotDefaultModValue ** {{{
   */
  hasNotDefaultModValue: function(modVal) {
    return !!modVal && modVal !== true;
  },/*}}}*/
  /** toModValue ** {{{
   */
  toModValue: function(modVal) {
    return typeof modVal === 'boolean' ? modVal : `'${modVal}'`;
  },/*}}}*/
  /** getEntityName ** {{{
   */
  getEntityName: function  ({ block, elem, mod={} }) {
    let entityName = block;
    if ( elem ) {
      entityName += this.elemDelim + elem;
    }
    if ( mod.name ) {
      entityName += this.modDelim + mod.name;
      if ( this.hasNotDefaultModValue(mod.val) ) {
        entityName += '_' + mod.val;
      }
    }
    return entityName;
  },/*}}}*/
  /** getObjectName ** {{{
   */
  getObjectName: function  ({ block, elem, mod={} }) {
    let entityName = block; // this.capitalize(block);
    if ( elem ) {
      entityName += this.capitalize(elem);
    }
    if ( mod.name ) {
      entityName += this.capitalize(mod.name);
      if ( mod.val && mod.val !== true ) {
        entityName += this.capitalize(mod.val);
      }
    }
    return entityName;
  },/*}}}*/

  /** getHeader ** {{{ Module header...
   */
  getHeader: function ({ block, elem, mod={} }) {

    const dateStr = require('dateformat')(new Date(), 'yyyy.mm.dd HH:MM');

    const entityName = this.getEntityName({ block, elem, mod });
    // const objectName = this.getObjectName({ block, elem, mod });
    // const jsEntityName = entityName.replace(/-/g, '__');
    // const jsEntityShortName = ( jsEntityName.length > block.length ) ? jsEntityName.substr(block.length) : jsEntityName;

    const at = '@';

    return `/**
 * ${at}module ${entityName}
 * ${at}author lilliputten <lilliputten@yandex.ru>
 * ${at}since ${dateStr}
 * ${at}version ${dateStr}
 */`;

  },/*}}}*/

}, config);


bemhtml.js
/* eslint-env es6, node, commonjs */
/**
 * @module bemhtml template
 * @author lilliputten <lilliputten@yandex.ru>
 * @since 2018.09.29, 23:06
 * @version 2018.09.29, 23:06
 */

const utils = require('./utils');

const extraHeaders = `// vim: ft=javascript
/* globals config, helpers, addAttrs, addElemMods, addJs, addMix, addMods, appendContent, apply, applyCtx, applyNext, attrs, bem, block, cls, content, def, elem, elemMod, elemMods, extend, js, local, match, mix, mod, mode, mods, oninit, prependContent, replace, tag, wrap, xjstOptions */`;

/** elemTmpl ** {{{
 */
const elemTmpl = (block, elem) => {

  const header = utils.getHeader({block, elem});
  // const objectName = utils.getObjectName({ block, elem });
  const entityName = utils.getEntityName({ block, elem });

  /** Content... {{{*/
  return `${extraHeaders}
${header}

block('${block}').elem('${elem}')(

  // def()(function(){
  //   this.escapeContent = false;
  //   return applyNext();
  // }),
  //
  // content()(function(){
  //   return [
  //     '${entityName}',
  //   ];
  // }),
  //
  // tag()()

);
`;/* ...Content }}}*/

};/*}}}*/

/** elemModTmpl ** {{{
 */
const elemModTmpl = (block, elem, mod) => {

  const { name: modName, val: modVal } = mod;
  const header = utils.getHeader({block, elem, mod});
  // const objectName = utils.getObjectName({ block, elem, mod });
  const entityName = utils.getEntityName({ block, elem, mod });
  // const blockElemName = block + utils.capitalize(elem);
  // const blockElemModName = blockElemName + utils.capitalize(modName) + utils.capitalize(utils.getModValString(modVal));

  /** Content... {{{*/
  return `${extraHeaders}
${header}

block('${block}').elem('${elem}').elemMod('${modName}', ${utils.toModValue(modVal)})(

  // def()(function(){
  //   this.escapeContent = false;
  //   return applyNext();
  // }),
  //
  // content()(function(){
  //   return [
  //     '${entityName}',
  //   ];
  // }),
  //
  // tag()()

);
`;/* ...Content }}}*/


};/*}}}*/

/** blockTmpl ** {{{
 */
const blockTmpl = (block) => {

  const header = utils.getHeader({block});
  const entityName = utils.getEntityName({ block });
  // const objectName = utils.getObjectName({ block });

  /** Content... {{{*/
  return `${extraHeaders}
${header}

block('${block}')(

  // def()(function(){
  //   this.escapeContent = false;
  //   return applyNext();
  // }),
  //
  // content()(function(){
  //   return [
  //     '${entityName}',
  //   ];
  // }),
  //
  // tag()()

);
`;/* ...Content }}}*/

};/*}}}*/

/** blockModTmpl ** {{{
 */
const blockModTmpl = (block, mod) => {

  const { name: modName, val: modVal } = mod;
  const header = utils.getHeader({block, mod});
  // const parentObjectName = utils.getObjectName({ block });
  // const objectName = utils.getObjectName({ block, mod });
  const entityName = utils.getEntityName({ block, mod });

  /** Content... {{{*/
  return `${extraHeaders}
${header}

block('${block}').mod('${modName}', ${utils.toModValue(modVal)})(

  // def()(function(){
  //   this.escapeContent = false;
  //   return applyNext();
  // }),
  //
  // content()(function(){
  //   return [
  //     '${entityName}',
  //   ];
  // }),
  //
  // tag()()

);
`;/* ...Content }}}*/

};/*}}}*/

/** module.exports ** {{{
 */
module.exports = ({ block, elem, mod={} }) => {

  const { name: modName } = mod;

  // TODO: elemModTmpl?
  if (block && elem && modName) {
    return elemModTmpl(block, elem, mod);
  }
  else if (block && elem) {
    return elemTmpl(block, elem);
  }
  else if (block && modName) {
    return blockModTmpl(block, mod);
  }
  else if (block) {
    return blockTmpl(block);
  }

};/*}}}*/


bemjson.js
/* eslint-env es6, node, commonjs */
/* eslint quotes:0, no-unused-vars:0 */
'use strict';
/**
 * @module bemjson template
 * @author lilliputten <lilliputten@yandex.ru>
 * @since 2018.09.29, 23:06
 * @version 2018.09.29, 23:06
 */

const utils = require('./utils');

/** module.exports ** {{{
 */
module.exports = function (entity, naming) {

  const entityName = utils.getEntityName({ block: entity });
  const header = utils.getHeader({block: entity});

  /** Content ... {{{*/
  return `/* eslint-env es6, node, commonjs */
// vim: ft=javascript
${header}

const

  Env = require('../../Env.js'),

  pageId = Env.getPageId(__dirname),

  // config = Env.config,

  /** pageContent ** {{{ Page content
   */
  pageContent = {
    block: 'test',
    cls: 'container',
    content: [
      '${entityName} content',
    ],
  },/*}}}*/

  /** bemjson ** {{{ Entire page
   */
  bemjson = Env.page({
    pageId,
    content: pageContent,
  })/*}}}*/

;

module.exports = bemjson; // Result
`;/* ...Content }}}*/

};/*}}}*/


js.js
/* eslint-env es6, node, commonjs */
/* eslint-disable semi */
/**
 * @module js template
 * @author lilliputten <lilliputten@yandex.ru>
 * @since 2018.09.29, 23:06
 * @version 2018.09.29, 23:06
 */

const utils = require('./utils');

/** elemTmpl ** {{{
 */
const elemTmpl = (block, elem) => {

  const header = utils.getHeader({block, elem});
  // const objectName = utils.getObjectName({ block, elem });
  const entityName = utils.getEntityName({ block, elem });

  /** Content... {{{*/
  return `${header}

modules.define('${block}__${elem}', [
  'i-bem-dom',
  // 'config',
  // 'vow',
],
function define(provide,
  bemDom,
  // config,
  // vow,
// eslint-disable-next-line no-unused-vars
__BASE) {

  /**
   * @exports
   * @class ${entityName}
   * @bem
   */
  var _${entityName}_proto = /** @lends ${entityName}_prototype */ {

    // /** _getDefaultParams ** {{{ */
    // _getDefaultParams: function() {
    //   // var params = this.__base(); // NOTE: For \`Object.assign\`
    //   return {
    //
    //     // paramName: 'value',
    //
    //   };
    // },/*}}}*/
    //
    // /** _${entityName}_init ** {{{ Module init... */
    // _${entityName}_init: function() {
    // },/*}}}*/
    //
    // /** onSetMod ** {{{ Modifiers... */
    // onSetMod: {
    //
    //   /** js ** {{{ JS lifecycle... */
    //   js: {
    //     inited: function() {
    //       this.__base.apply(this, arguments);
    //       this._${entityName}_init && this._${entityName}_init();
    //     },
    //   },/*}}}*/
    //
    // },/*}}}*/

  };

  // /** _${entityName}_static ** {{{ Static properties... */
  // var _${entityName}_static = /** @lends ${entityName} */ {
  //
  //   // lazyInit : true,
  //
  //   /** onInit ** {{{ Static init...
  //    */
  //   onInit: function() {
  //
  //     // var proto = this.prototype;
  //     // this._events({ block : Button, modName : 'action', modVal : 'itemClose' }).on('click', proto.onItemCloseButtonClick);
  //
  //     return this.__base.apply(this, arguments);
  //
  //   }/*}}}*/
  //
  // };/*}}}*/

  // Provide element...
  provide(bemDom.declElem('${block}', '${elem}', _${entityName}_proto));

}); // Module end
`;/* ...Content }}}*/

};/*}}}*/

/** elemModTmpl ** {{{
 */
const elemModTmpl = (block, elem, mod) => {

  const { name: modName, val: modVal } = mod;
  const header = utils.getHeader({block, elem, mod});
  // const objectName = utils.getObjectName({ block, elem, mod });
  const entityName = utils.getEntityName({ block, elem, mod });
  // const blockElemName = block + utils.capitalize(elem);
  // const blockElemModName = blockElemName + utils.capitalize(modName) + utils.capitalize(utils.getModValString(modVal));

  /** Content... {{{*/
  return `${header}

modules.define('${block}__${elem}', [
  // 'i-bem-dom',
  // 'config',
  // 'vow',
],
function define(provide,
  // bemDom,
  // config,
  // vow,
${block}__${elem}) {

  /**
   * @exports
   * @class ${entityName}
   * @bem
   */
  var _${entityName}_proto = /** @lends ${entityName}_prototype */ {

    // /** _getDefaultParams ** {{{ */
    // _getDefaultParams: function() {
    //   // var params = this.__base(); // NOTE: For \`Object.assign\`
    //   return {
    //
    //     // paramName: 'value',
    //
    //   };
    // },/*}}}*/
    //
    // /** _${entityName}_init ** {{{ Module init... */
    // _${entityName}_init: function() {
    // },/*}}}*/
    //
    // /** onSetMod ** {{{ Modifiers... */
    // onSetMod: {
    //
    //   /** js ** {{{ JS lifecycle... */
    //   js: {
    //     inited: function() {
    //       this.__base.apply(this, arguments);
    //       this._${entityName}_init && this._${entityName}_init();
    //     },
    //   },/*}}}*/
    //
    // },/*}}}*/

  };

  // /** _${entityName}_static ** {{{ Static properties... */
  // var _${entityName}_static = /** @lends ${entityName} */ {
  //
  //   // lazyInit : true,
  //
  //   /** onInit ** {{{ Static init...
  //    */
  //   onInit: function() {
  //
  //     // var proto = this.prototype;
  //     // this._events({ block : Button, modName : 'action', modVal : 'itemClose' }).on('click', proto.onItemCloseButtonClick);
  //
  //     return this.__base.apply(this, arguments);
  //
  //   }/*}}}*/
  //
  // };/*}}}*/

  // Provide element modifier
  provide(${block}__${elem}.declMod({modName: '${modName}', modVal: ${utils.toModValue(modVal)}}, _${entityName}_proto));

}); // Module end
`;/* ...Content }}}*/

};/*}}}*/

/** blockTmpl ** {{{
 */
const blockTmpl = (block) => {

  const header = utils.getHeader({block});
  const entityName = utils.getEntityName({ block });
  // const objectName = utils.getObjectName({ block });

  /** Content... {{{*/
  return `${header}

modules.define('${entityName}', [
  'i-bem-dom',
  // 'config',
  // 'vow',
],
function define(provide,
  bemDom,
  // config,
  // vow,
// eslint-disable-next-line no-unused-vars
__BASE) {

  /**
   * @exports
   * @class ${entityName}
   * @bem
   */
  var _${entityName}_proto = /* @lends ${entityName}.prototype */ {

    // /** _getDefaultParams ** {{{ */
    // _getDefaultParams: function() {
    //   // var params = this.__base(); // NOTE: For \`Object.assign\`
    //   return {
    //
    //     // paramName: 'value',
    //
    //   };
    // },/*}}}*/
    //
    // /** _${entityName}_init ** {{{ Block init...
    //  */
    // _${entityName}_init: function() {
    // },/*}}}*/
    //
    // /** onSetMod ** {{{ Modifiers... */
    // onSetMod: {
    //
    //   /** js ** {{{ JS lifecycle... */
    //   js: {
    //     inited: function() {
    //       this.__base.apply(this, arguments);
    //       this._${entityName}_init && this._${entityName}_init();
    //     },
    //   },/*}}}*/
    //
    // },/*}}}*/

  };

  // /** _${entityName}_static ** {{{ Static properties... */
  // var _${entityName}_static = /** @lends ${entityName} */ {
  //
  //   // lazyInit : true,
  //
  //   /** onInit ** {{{ Static init...
  //    */
  //   onInit: function() {
  //
  //     // var proto = this.prototype;
  //     // this._events({ block : Button, modName : 'action', modVal : 'itemClose' }).on('click', proto.onItemCloseButtonClick);
  //
  //     return this.__base.apply(this, arguments);
  //
  //   }/*}}}*/
  //
  // };/*}}}*/

  // Provide block...
  provide(bemDom.declBlock(this.name, _${entityName}_proto));

}); // Module end
`;/* ...Content }}}*/

};/*}}}*/

/** blockModTmpl ** {{{
 */
const blockModTmpl = (block, mod) => {

  const { name: modName, val: modVal } = mod;
  const header = utils.getHeader({block, mod});
  // const parentObjectName = utils.getObjectName({ block });
  // const objectName = utils.getObjectName({ block, mod });
  const entityName = utils.getEntityName({ block, mod });

  /** Content... {{{*/
  return `${header}

modules.define('${block}', [
  'i-bem-dom',
  // 'config',
],
  // 'vow',
function define(provide,
  bemDom,
  // config,
${block}) {
  // vow,

  /**
   * @exports
   * @class ${entityName}
   * @bem
   */
  var _${entityName}_proto = /** @lends ${entityName}_prototype */ {

    // /** _getDefaultParams ** {{{ */
    // _getDefaultParams: function() {
    //   // var params = this.__base(); // NOTE: For \`Object.assign\`
    //   return {
    //
    //     // paramName: 'value',
    //
    //   };
    // },/*}}}*/
    //
    // /** _${entityName}_init ** {{{ Module init... */
    // _${entityName}_init: function() {
    // },/*}}}*/
    //
    // /** onSetMod ** {{{ Modifiers... */
    // onSetMod: {
    //
    //   /** js ** {{{ JS lifecycle... */
    //   js: {
    //     inited: function() {
    //       this.__base.apply(this, arguments);
    //       this._${entityName}_init && this._${entityName}_init();
    //     },
    //   },/*}}}*/
    //
    // },/*}}}*/

  };

  // /** _${entityName}_static ** {{{ Static properties... */
  // var _${entityName}_static = /** @lends ${entityName} */ {
  //
  //   // lazyInit : true,
  //
  //   /** onInit ** {{{ Static init...
  //    */
  //   onInit: function() {
  //
  //     // var proto = this.prototype;
  //     // this._events({ block : Button, modName : 'action', modVal : 'itemClose' }).on('click', proto.onItemCloseButtonClick);
  //
  //     return this.__base.apply(this, arguments);
  //
  //   }/*}}}*/
  //
  // };/*}}}*/

  // Provide block modifier
  provide(${block}.declMod({modName: '${modName}', modVal: ${utils.toModValue(modVal)}}, _${entityName}_proto));

}); // Module end
`;/* ...Content }}}*/

}/*}}}*/

/** module.exports ** {{{
 */
module.exports = ({ block, elem, mod={} }) => {

  const { name: modName } = mod;

  // TODO: elemModTmpl?
  if (block && elem && modName) {
    return elemModTmpl(block, elem, mod);
  }
  else if (block && elem) {
    return elemTmpl(block, elem);
  }
  else if (block && modName) {
    return blockModTmpl(block, mod);
  }
  else if (block) {
    return blockTmpl(block);
  }

}/*}}}*/
